Intersect.Client/Entities/Entity.cs

```diff
@@ -1931,10 +1999,47 @@ protected byte DirectionToTarget(Entity en)
 
             var yOffset = originY - targetY;
             var xOffset = originX - targetX;
-            var xDir = xOffset == 0 ? newDir : (xOffset < 0 ? (byte)Directions.Right : (byte)Directions.Left);
-            var yDir = yOffset == 0 ? newDir : (yOffset < 0 ? (byte)Directions.Down : (byte)Directions.Up);
 
-            newDir = Math.Abs(yOffset) > Math.Abs(xOffset) ? yDir : xDir;
+            newDir = Options.Instance.MapOpts.EnableDiagonalMovement
+                ? CrossAndDiagonalDirections(newDir, xOffset, yOffset)
+                : CrossDirections(newDir, xOffset, yOffset);
+
+            Directions CrossDirections(Directions direction, int xOffset4, int yOffset4)
+            {
+                var xDir = xOffset4 == 0 ? direction : (xOffset4 < 0 ? Directions.Right : Directions.Left);
+                var yDir = yOffset4 == 0 ? direction : (yOffset4 < 0 ? Directions.Down : Directions.Up);
+                direction = Math.Abs(yOffset4) > Math.Abs(xOffset4) ? yDir : xDir;
+
+                return direction;
+            }
+
+            Directions CrossAndDiagonalDirections(Directions direction, int xOffset8, int yOffset8)
+            {
+                if (xOffset8 == 0)
+                {
+                    if (yOffset8 == 0)
+                    {
+                        return direction;
+                    }
+
+                    direction = yOffset8 > 0 ? Directions.Up : Directions.Down;
+                }
+                else if (yOffset8 == 0)
+                {
+                    direction = xOffset8 > 0 ? Directions.Left : Directions.Right;
+                }
+                else
+                {
+                    int xSign = Math.Sign(xOffset8);
+                    int ySign = Math.Sign(yOffset8);
+
+                    direction = xSign > 0
+                        ? (ySign > 0 ? Directions.UpLeft : Directions.DownLeft)
+                        : (ySign > 0 ? Directions.UpRight : Directions.DownRight);
+                }
+
+                return direction;
+            }
```

Comment:
I have no idea WTF the code was doing before this, and I certainly have no idea now. Please explain what the old code did and why the new code needs two local methods (instead of an if statement).

Reply:
👍🏽  ; How about :

```
protected Direction DirectionToTarget(Entity en)
        {
            Direction newDir = Dir;

            if (en == null)
            {
                return newDir;
            }

            int originY = Y;
            int originX = X;
            int targetY = en.Y;
            int targetX = en.X;

            // Calculate Y and X offset between origin and target if they're not on the same map instance.
            if (en.MapInstance.Id != MapInstance.Id)
            {
                if (en.MapInstance.GridY < MapInstance.GridY)
                {
                    originY += Options.MapHeight - 1;
                }
                else if (en.MapInstance.GridY > MapInstance.GridY)
                {
                    targetY += Options.MapHeight - 1;
                }

                if (en.MapInstance.GridX < MapInstance.GridX)
                {
                    originX += Options.MapWidth - 1;
                }
                else if (en.MapInstance.GridX > MapInstance.GridX)
                {
                    targetX += (Options.MapWidth - 1);
                }
            }

            int yDiff = originY - targetY;
            int xDiff = originX - targetX;

            // If diagonal movement is not enabled, direction is determined by the largest offset (X or Y).
            if (!Options.Instance.MapOpts.EnableDiagonalMovement)
            {
                return Math.Abs(yDiff) > Math.Abs(xDiff)
                    ? (yDiff > 0 ? Direction.Up : Direction.Down)
                    : (xDiff > 0 ? Direction.Left : Direction.Right);
            }

            // Diagonal movement is enabled !
            // If X offset is 0, direction is determined by Y offset.
            if (xDiff == 0)
            {
                return yDiff > 0 ? Direction.Up : Direction.Down;
            }

            // If Y offset is 0, direction is determined by X offset.
            if (yDiff == 0)
            {
                return xDiff > 0 ? Direction.Left : Direction.Right;
            }

            // If both X and Y offset are non-zero, direction is determined by both offsets.
            int xSign = Math.Sign(xDiff);
            int ySign = Math.Sign(yDiff);

            return xSign > 0
                ? (ySign > 0 ? Direction.UpLeft : Direction.DownLeft)
                : (ySign > 0 ? Direction.UpRight : Direction.DownRight);
        }
```

Reply:
![image](https://user-images.githubusercontent.com/1476550/217388155-e7cc1c97-5d0c-4946-9103-4e32921d3a3a.png)
this right here is almost exactly the same code that was used for when diagonal is disabled.

You can simplify it by changing the second if statement to also have `|| !Diagonal`

That said, I would change the order, it should check if yDiff == 0 first, and then check xDiff (or diagonal off)

Reply:
I would also split the ternary into if statements 
![image](https://user-images.githubusercontent.com/1476550/217389661-b6073be4-45fd-43b2-8dc7-24cd7a25b2d0.png)

Reply:
```
        protected Direction DirectionToTarget(Entity en)
        {
            Direction newDir = Dir;

            if (en == null)
            {
                return newDir;
            }

            int originY = Y;
            int originX = X;
            int targetY = en.Y;
            int targetX = en.X;

            // Calculate Y and X offset between origin and target if they're not on the same map instance.
            if (en.MapInstance.Id != MapInstance.Id)
            {
                if (en.MapInstance.GridY < MapInstance.GridY)
                {
                    originY += Options.MapHeight - 1;
                }
                else if (en.MapInstance.GridY > MapInstance.GridY)
                {
                    targetY += Options.MapHeight - 1;
                }

                if (en.MapInstance.GridX < MapInstance.GridX)
                {
                    originX += Options.MapWidth - 1;
                }
                else if (en.MapInstance.GridX > MapInstance.GridX)
                {
                    targetX += (Options.MapWidth - 1);
                }
            }

            // Calculate the offset between origin and target along both of their axis.
            int yDiff = originY - targetY;
            int xDiff = originX - targetX;

            // If Y offset is 0, direction is determined by X offset.
            if (yDiff == 0)
            {
                return xDiff > 0 ? Direction.Left : Direction.Right;
            }

            // If X offset is 0 or If diagonal movement is disabled, direction is determined by Y offset.
            if (xDiff == 0 || !Options.Instance.MapOpts.EnableDiagonalMovement)
            {
                return yDiff > 0 ? Direction.Up : Direction.Down;
            }

            // If both X and Y offset are non-zero, direction is determined by both offsets.
            int xSign = Math.Sign(xDiff);
            int ySign = Math.Sign(yDiff);

            return xSign > 0
                ? (ySign > 0 ? Direction.UpLeft : Direction.DownLeft)
                : (ySign > 0 ? Direction.UpRight : Direction.DownRight);
        }
```